File: python/sorting/Cycle Sort.py
def cycle_sort(num_list):
    "Sort a list in place and return the number of writes."
    writes = 0
 
    for list_start, item in enumerate(num_list):
 
        pos = list_start
        for item2 in num_list[list_start + 1:]:
            if item2 < item:
                pos += 1

        if pos == list_start:
            continue
 
        while item == num_list[pos]:
            pos += 1
        num_list[pos], item = item, num_list[pos]
        writes += 1
 
        while pos != list_start:
 
            pos = list_start
            for item2 in num_list[list_start + 1:]:
                if item2 < item:
                    pos += 1
 
            while item == num_list[pos]:
                pos += 1
            num_list[pos], item = item, num_list[pos]
            writes += 1
 
    return writes
 
 
num_list = [75, 16, 55, 19, 48, 14, 2, 61, 22, 100]
print("Before: ", num_list)
cycle_sort(num_list)
print("After:  ", num_list)

File: java/sorting/cycle-sort.java
import java.util.*;
import java.lang.*;

class GFG {
	
	public static void cycleSort(int arr[], int n)
	{
		
		int writes = 0;

		

		
		
		for (int cycle_start = 0; cycle_start <= n - 2; cycle_start++) {
			
			int item = arr[cycle_start];

			
			
			int pos = cycle_start;
			for (int i = cycle_start + 1; i < n; i++)
				if (arr[i] < item)
					pos++;

			
			if (pos == cycle_start)
				continue;

			
			while (item == arr[pos])
				pos += 1;

			
			if (pos != cycle_start) {
				int temp = item;
				item = arr[pos];
				arr[pos] = temp;
				writes++;
			}

			
			while (pos != cycle_start) {
				pos = cycle_start;

				
				for (int i = cycle_start + 1; i < n; i++)
					if (arr[i] < item)
						pos += 1;

				
				while (item == arr[pos])
					pos += 1;

				
				if (item != arr[pos]) {
					int temp = item;
					item = arr[pos];
					arr[pos] = temp;
					writes++;
				}
			}
		}
	}

	
	public static void main(String[] args)
	{
		int arr[] = { 1, 8, 3, 9, 10, 10, 2, 4 };
		int n = arr.length;
		cycleSort(arr, n);

		System.out.println("After sort : ");
		for (int i = 0; i < n; i++)
			System.out.print(arr[i] + " ");
	}
}
